File: shell-ios/CloudGuard/BridgeRouter.swift
import Foundation
import UIKit

/// 桥接路由器：把 JS 的 method 分发到各原生能力模块（对应 Android 端 BridgeRouter）。
///
/// 与 utils/h5-native-bridge.js 的方法清单严格对应：
///   device.getInfo / permission.request / scan.qrcode
///   ble.* / image.* / share.* / app.* / ad.* / pay.* / http.* / miniprogram.*
///
/// 所有处理统一切到主线程执行。
final class BridgeRouter {

    private weak var host: WebViewController?
    private let channel: JsChannel

    private let ble: BleManager
    private let media: MediaManager
    private let share: ShareManager
    private let update: UpdateManager
    private let ad: AdBridge
    private let pay: PayBridge
    private let http: HttpManager
    private let miniProgram: MiniProgramBridge

    init(host: WebViewController, channel: JsChannel) {
        self.host = host
        self.channel = channel
        self.ble = BleManager(host: host, channel: channel)
        self.media = MediaManager(host: host, channel: channel)
        self.share = ShareManager(host: host, channel: channel)
        self.update = UpdateManager(host: host, channel: channel)
        self.ad = AdBridge(host: host, channel: channel)
        self.pay = PayBridge(channel: channel)
        self.http = HttpManager(channel: channel)
        self.miniProgram = MiniProgramBridge(channel: channel)
    }

    func dispatch(method: String, params: [String: Any], callbackId: String) {
        DispatchQueue.main.async { [weak self] in
            self?.safeRoute(method: method, params: params, callbackId: callbackId)
        }
    }

    /// 页面销毁：释放各能力模块持有的资源（广告悬浮视图、支付队列观察者等）
    func dispose() {
        ad.dispose()
        pay.dispose()
    }

    private func safeRoute(method: String, params: [String: Any], callbackId: String) {
        do {
            try route(method: method, params: params, callbackId: callbackId)
        } catch {
            channel.resolveErr(callbackId, "bridge error: \(error.localizedDescription)")
        }
    }

    private func route(method: String, params: [String: Any], callbackId: String) throws {
        switch true {
        case method == "device.getInfo":
            channel.resolveOk(callbackId, deviceInfo())
        case method == "permission.request":
            handlePermission(params: params, callbackId: callbackId)
        case method == "scan.qrcode":
            host?.requestScan(callbackId: callbackId)
        case method.hasPrefix("ble."):
            ble.handle(method: method, params: params, callbackId: callbackId)
        case method.hasPrefix("image."):
            media.handle(method: method, params: params, callbackId: callbackId)
        case method.hasPrefix("share."):
            share.handle(method: method, params: params, callbackId: callbackId)
        case method.hasPrefix("app."):
            update.handle(method: method, params: params, callbackId: callbackId)
        case method.hasPrefix("ad."):
            ad.handle(method: method, params: params, callbackId: callbackId)
        case method.hasPrefix("pay."):
            pay.handle(method: method, params: params, callbackId: callbackId)
        case method.hasPrefix("miniprogram."):
            _ = miniProgram.handle(method: method, params: params, callbackId: callbackId)
        case method.hasPrefix("http."):
            http.handle(method: method, params: params, callbackId: callbackId)
        default:
            channel.resolveErr(callbackId, "unknown method: \(method)")
        }
    }

    // MARK: - 权限

    /// iOS 权限在首次使用时由系统弹窗；此处按请求清单返回 granted=true（真实授权结果由系统决定）。
    private func handlePermission(params: [String: Any], callbackId: String) {
        let perms = params["permissions"] as? [String] ?? []
        var result: [String: Any] = [:]
        for p in perms { result[p] = true }
        channel.resolveOk(callbackId, result)
    }

    // MARK: - 设备信息

    private func deviceInfo() -> [String: Any] {
        let info = Bundle.main.infoDictionary
        let version = info?["CFBundleShortVersionString"] as? String ?? ""
        let build = info?["CFBundleVersion"] as? String ?? ""
        return [
            "platform": "ios",
            "system": "iOS \(UIDevice.current.systemVersion)",
            "model": UIDevice.current.model,
            "brand": "Apple",
            "version": version,
            "buildNumber": build,
            "appId": Bundle.main.bundleIdentifier ?? ""
        ]
    }
}

File: shell-android/app/src/main/java/com/yefiot/cloudguard/bridge/BridgeRouter.kt
package com.yefiot.cloudguard.bridge

import android.os.Build
import android.os.Handler
import android.os.Looper
import com.yefiot.cloudguard.MainActivity
import com.yefiot.cloudguard.ad.AdBridge
import com.yefiot.cloudguard.ble.BleManager
import com.yefiot.cloudguard.http.HttpManager
import com.yefiot.cloudguard.media.MediaManager
import com.yefiot.cloudguard.miniprogram.MiniProgramBridge
import com.yefiot.cloudguard.pay.PayBridge
import com.yefiot.cloudguard.share.ShareManager
import com.yefiot.cloudguard.update.UpdateManager
import org.json.JSONArray
import org.json.JSONObject

/**
 * 桥接路由器：把 JS 的 method 分发到各原生能力模块。
 *
 * 与 utils/h5-native-bridge.js 的方法清单严格对应：
 *   device.getInfo / permission.request / scan.qrcode
 *   ble.* / image.* / share.* / app.* / ad.* / pay.* / miniprogram.*
 *
 * 所有处理统一切到主线程执行（Activity / WebView / 广告 SDK 均要求主线程）。
 */
class BridgeRouter(
    private val activity: MainActivity,
    private val channel: JsChannel
) {
    private val mainHandler = Handler(Looper.getMainLooper())

    private val bleManager = BleManager(activity, channel)
    private val mediaManager = MediaManager(activity, channel)
    private val shareManager = ShareManager(activity, channel)
    private val updateManager = UpdateManager(activity, channel)
    private val adBridge = AdBridge(activity, channel)
    private val httpManager = HttpManager(channel)
    private val payBridge = PayBridge(activity, channel)
    private val miniProgramBridge = MiniProgramBridge(activity, channel)

    /** 页面销毁时释放桥接资源（清 PayBridge 静态引用等，避免泄漏 WebView） */
    fun dispose() {
        payBridge.dispose()
    }

    fun dispatch(method: String, params: JSONObject, callbackId: String) {
        mainHandler.post { safeRoute(method, params, callbackId) }
    }

    private fun safeRoute(method: String, params: JSONObject, callbackId: String) {
        try {
            route(method, params, callbackId)
        } catch (t: Throwable) {
            channel.resolveErr(callbackId, "bridge error: ${t.message}")
        }
    }

    private fun route(method: String, params: JSONObject, callbackId: String) {
        when {
            method == "device.getInfo" -> channel.resolveOk(callbackId, deviceInfo())
            method == "permission.request" -> handlePermission(params, callbackId)
            method == "scan.qrcode" -> activity.requestScan(callbackId)
            method.startsWith("ble.") -> bleManager.handle(method, params, callbackId)
            method.startsWith("image.") -> mediaManager.handle(method, params, callbackId)
            method.startsWith("share.") -> shareManager.handle(method, params, callbackId)
            method.startsWith("app.") -> updateManager.handle(method, params, callbackId)
            method.startsWith("ad.") -> adBridge.handle(method, params, callbackId)
            method.startsWith("pay.") -> payBridge.handle(method, params, callbackId)
            method.startsWith("miniprogram.") -> miniProgramBridge.handle(method, params, callbackId)
            method.startsWith("http.") -> httpManager.handle(method, params, callbackId)
            else -> channel.resolveErr(callbackId, "unknown method: $method")
        }
    }

    private fun handlePermission(params: JSONObject, callbackId: String) {
        val arr = params.optJSONArray("permissions") ?: JSONArray()
        val perms = ArrayList<String>()
        for (i in 0 until arr.length()) {
            val p = arr.optString(i)
            if (p.isNotEmpty()) perms.add(p)
        }
        activity.requestPermissions(perms.toTypedArray()) { result ->
            val obj = JSONObject()
            result.forEach { (k, v) -> obj.put(k, v) }
            channel.resolveOk(callbackId, obj)
        }
    }

    private fun deviceInfo(): JSONObject {
        val pi = activity.packageManager.getPackageInfo(activity.packageName, 0)
        val obj = JSONObject()
        obj.put("platform", "android")
        obj.put("system", "Android ${Build.VERSION.RELEASE}")
        obj.put("model", Build.MODEL)
        obj.put("brand", Build.BRAND)
        obj.put("version", pi.versionName ?: "")
        obj.put(
            "buildNumber",
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) pi.longVersionCode.toString() else pi.versionCode.toString()
        )
        obj.put("appId", activity.packageName)
        return obj
    }
}
